### PR TITLE
arm64 mm: asid and tlb support

### PR DESCRIPTION
**Pull request by lubinszARM**
Some optimizations in this pr:
  1, Move ASID from TTBR0 to TTBR1
  2, tlb_flush_all

To-do: implementations for 
	__tlbi(aside1is, asid)
	__tlbi_user(aside1is, asid)

Signed-off-by: Bin Lu <bin.lu@arm.com>

### CODE REVIEW

**Comment by lubinszARM** [comment]
Related issue: https://github.com/google/gvisor/issues/3019

**Comment by lubinszARM** [comment]
Hi @avagin 
Any comments?
Thanks.

File: pkg/sentry/platform/ring0/entry_arm64.s [hunk 1]
```diff
@@ -327,7 +335,6 @@
 	SUB $16, RSP, RSP; \		// step1, save r18, r9 into kernel temporary stack.
 	STP (RSV_REG, RSV_REG_APP), 16*0(RSP); \
 	WORD $0xd538d092; \    //MRS   TPIDR_EL1, R18, step2, switch user pagetable.
-	SWITCH_TO_KVM_PAGETABLE(RSV_REG); \
 	WORD $0xd538d092; \    //MRS   TPIDR_EL1, R18
```
Comment (nixprime) [inline pkg/sentry/platform/ring0/entry_arm64.s:338]: Isn't this second MRS redundant now?

File: pkg/sentry/platform/ring0/entry_arm64.s [hunk 2]
```diff
@@ -300,17 +300,25 @@
...
 #define SWITCH_TO_APP_PAGETABLE(from) \
-	MOVD CPU_TTBR0_APP(from), RSV_REG; \
-	WORD $0xd5182012; \	//        MSR R18, TTBR0_EL1
 	ISB $15; \
-	DSB $15;
+	MRS TTBR1_EL1, R0; \
+	MOVD CPU_APP_ASID(from), R1; \
+	BFI $48, R1, $16, R0; \
+	MSR R0, TTBR1_EL1; \
+	ISB $15; \
```
Comment (nixprime) [inline pkg/sentry/platform/ring0/entry_arm64.s:308]: Can you clarify why two ISBs are needed in this macro (and `SWITCH_TO_KVM_PAGETABLE`), before both the stores to TTBR1_EL1 and TTBR0_EL1 respectively?

File: pkg/sentry/platform/ring0/lib_arm64.s [hunk 3]
```diff
@@ -15,6 +15,13 @@
 #include "funcdata.h"
 #include "textflag.h"
 
+TEXT ·FlushTlbAll(SB),NOSPLIT,$0
+	DSB $10
+	WORD $0xd508831f
```
Comment (nixprime) [inline pkg/sentry/platform/ring0/lib_arm64.s:20]: Can you add a comment indicating what instruction this encodes? (IIUC it's TLBI VMALLE1IS)

File: pkg/sentry/platform/ring0/entry_arm64.s [hunk 4]
```diff
@@ -357,6 +364,14 @@
 	MOVD R4, CPU_REGISTERS+PTRACE_SP(RSV_REG); \
 	LOAD_KERNEL_STACK(RSV_REG);  // Load the temporary stack.
 
+// storeAppASID writes the application's asid value.
+TEXT ·storeAppASID(SB),NOSPLIT,$0-8
+	MOVD asid+0(FP), R1
+	MRS  TPIDR_EL1, RSV_REG
+	MOVD R1, CPU_APP_ASID(RSV_REG)
+	DSB $7
```
Comment (nixprime) [inline pkg/sentry/platform/ring0/entry_arm64.s:372]: Why is DSB NSH needed here?

File: pkg/sentry/platform/ring0/entry_arm64.s (hunk 2 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/entry_arm64.s:307]: Hi @nixprime 
I have removed the ISB before the store.
I think one ISB after  'MSR R0, TTBR1_EL1' is enough to ensure ensure that possible change is visible by the core.
Thanks for your reminder.

File: pkg/sentry/platform/ring0/entry_arm64.s (hunk 1 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/entry_arm64.s:335]: Done.

File: pkg/sentry/platform/ring0/entry_arm64.s (hunk 4 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/entry_arm64.s]: 'dsb nsh' was removed.
Thanks.

File: pkg/sentry/platform/ring0/lib_arm64.s (hunk 3 above)
Comment (lubinszARM) [reply pkg/sentry/platform/ring0/lib_arm64.s]: Done.

**Pull request merged**